### PR TITLE
Pytorch sparse workload

### PR DESCRIPTION
**Pull request by zzy82518996**
This pull request aims to merge all the exetend HB pytorch operators for sparse workloads into the master, changes includes follows:
1. Low level copy support for sparse tensor
2. Element-wise sparse tensor and dense tensor
3. SpMV operator (torch.mv): 
4. SpMM operator (torch.mm)
5. element-wise max for float data type
6. element-wise remainder for integer data type
7. tensor indexing operation
8. torch.sign
9. torch.random
10. torch.cat (extended from master)
11. placeholder for sparse-sparse add and sparse-sparse mul
12. placeholder for spmvxcel and spmmxcel, (cann't be correctly executed under emulation mode)
12. Optimization for indexing non-contiguous tensor (remove div and mod )
Changes mainly focus inside hb-pytorch/hammerblade/torch/kernel/, hb-pytorch/aten/src/ATen/native/hammerblade and hb-pytorch/aten/src/ATen/native/sparse/hammerblade

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -139,7 +139,7 @@ option(USE_HB "Use HammerBlade" ON)
 option(USE_HB_EMUL "Use HammerBlade Emulation Layer" OFF)
 option(PROFILE_ATEN "Turn on ATen kernel profiling" ON)
 cmake_dependent_option(
-  HB_REDISPATCH "Enable conditional redispatch to HB" ON
+  HB_REDISPATCH "Enable conditional redispatch to HB" OFF
```
Comment (yodada) [inline CMakeLists.txt]: looks like this change is causing some of the failures we have seen on Travis

Comment (zzy82518996) [reply CMakeLists.txt]: Good point ! Thanks !

File: aten/src/ATen/function_wrapper.py [hunk 2]
```diff
@@ -2153,6 +2153,7 @@ def process_redispatch_cpu_to_hb(option):
 
         # get alternative dispatching dst
         dispatch = option['type_method_definition_dispatch']
+        print(dispatch)
```
Comment (yodada) [inline aten/src/ATen/function_wrapper.py]: not critical
but here is a leftover ;)

File: aten/src/ATen/function_wrapper.py [hunk 3]
```diff
@@ -2257,10 +2426,22 @@ def process_native(option):
...
                     else:
                         option['log_unimpl'] = ''
+                    if backend == 'SparseCPU' and 'SparseHammerBlade' not in option['backend_types']:
+                        option['log_unimpl'] = 'if (c10::probe::hb_profiler_is_in_roi()) {\n'
+                        option['log_unimpl'] += '  c10::probe::log_unimpl_kernel("aten::{0}");\n'.format(option['api_name'])
+                        option['log_unimpl'] += '}'
+                    else:
+                        option['log_unimpl'] = ''
                     if backend == 'CPU' and 'HammerBlade' in option['backend_types']:
+                        print("Enter CPU backend")
```
Comment (yodada) [inline aten/src/ATen/function_wrapper.py]: same here and below ;)

File: aten/src/ATen/native/TensorIterator.cpp [hunk 4]
```diff
@@ -1080,7 +1081,7 @@ void DimCounter::increment(const std::array<int64_t, 2>& step) {
 }
 
 std::array<int64_t, 2> DimCounter::max_2d_step() const {
-  int64_t step0 = std::min(shape[0] - values[0], range.end - offset);
+  int64_t step0 = std::min(shape[0] - values[0], range.end - offset);
```
Comment (yodada) [inline aten/src/ATen/native/TensorIterator.cpp:1083]: hmmm these seem not really changing anything? maybe checkout from master so it's not in the changeset?

File: aten/src/ATen/native/hammerblade/OffloadUtils.cpp [hunk 5]
```diff
@@ -96,9 +96,9 @@ eva_t create_device_tensor(const Tensor& tensor_arg,
     tensor = tensor_arg;
   }
 
-  if(!tensor.is_contiguous()) {
```
Comment (yodada) [inline aten/src/ATen/native/hammerblade/OffloadUtils.cpp]: I kind of want to have this warning

File: c10/hammerblade/HammerBladeFunctions.cpp [hunk 6]
```diff
@@ -238,7 +238,7 @@ void offload_kernel(const char* kernel, std::vector<eva_t> args) {
 #endif
 
   // debug
-  std::cerr << kernel << " finished -- abs cycle = " << abs_cycle << std::endl;
+//  std::cerr << kernel << " finished -- abs cycle = " << abs_cycle << std::endl;
```
Comment (yodada) [inline c10/hammerblade/HammerBladeFunctions.cpp:241]: oops this one needs to be on

File: hammerblade/torch/dramsim3epoch.json [hunk 7]
```diff
@@ -0,0 +1,8 @@
...tency[90-99]":3,"num_act_cmds":1329,"num_cycles":1000000,"num_ondemand_pres":65,"num_pre_cmds":1329,"num_read_cmds":4484,"num_read_row_hits":3359,"num_reads_done":4484,"num_ref_cmds":256,"num_refb_cmds":0,"num_srefe_cmds":0,"num_srefx_cmds":0,"num_write_buf_hits":0,"num_write_cmds":4415,"num_write_row_hits":4224,"num_writes_done":4420,"pre_stb_energy":{"0":29255808.0},"rank_active_cycles":{"0":390504},"read_energy":1802568.0,"read_latency[-0]":0,"read_latency[0-19]":0,"read_latency[100-119]":0,"read_latency[120-139]":0,"read_latency[140-159]":1,"read_latency[160-179]":3,"read_latency[180-199]":16,"read_latency[20-39]":4292,"read_latency[200-]":168,"read_latency[40-59]":0,"read_latency[60-79]":4,"read_latency[80-99]":0,"ref_energy":15575040.0,"refb_energy":-0.0,"sref_cycles":{"0":0},"sref_energy":{"0":0.0},"tag":0,"total_energy":75864702.0,"write_energy":2357610.0,"write_latency[-0]":0,"write_latency[0-19]":482,"write_latency[100-119]":27,"write_latency[120-139]":35,"write_latency[140-159]":40,"write_latency[160-179]":48,"write_latency[180-199]":31,"write_latency[20-39]":1636,"write_latency[200-]":216,"write_latency[40-59]":1728,"write_latency[60-79]":129,"write_latency[80-99]":43},
```
Comment (yodada) [inline hammerblade/torch/dramsim3epoch.json]: ahhh this shouldnt be here?

File: hammerblade/torch/kernel/kernel_threshold.cpp [hunk 8]
```diff
@@ -30,6 +31,7 @@ extern "C" {
         } else {
           return other;
         }
+        printf("self and other are : \n", self, other);
```
Comment (yodada) [inline hammerblade/torch/kernel/kernel_threshold.cpp]: cant have printf in the kernels -- cant build cosim ;)

**Review by yodada — COMMENTED** [review]
Two general comments:
 - there are a few files that shouldnt be staged -- out.std, vcache_operation_trace.csv, and vcache_stats.csv
 - it looks like the hacky recsys embedding thing is included in this PR. I'm not sure we want to have it in master branch

File: c10/hammerblade/HammerBladeFunctions.cpp (hunk 6 above)
Comment (zzy82518996) [reply c10/hammerblade/HammerBladeFunctions.cpp:241]: This is the place lead to travis issue, should I still turn it on ?

File: aten/src/ATen/native/TensorIterator.cpp (hunk 4 above)
Comment (zzy82518996) [reply aten/src/ATen/native/TensorIterator.cpp:1083]: There is one blank line 806. I have fixed this issue.

File: aten/src/ATen/function_wrapper.py (hunk 3 above)
Comment (zzy82518996) [reply aten/src/ATen/function_wrapper.py]: Done.

File: hammerblade/torch/kernel/kernel_threshold.cpp (hunk 8 above)
Comment (zzy82518996) [reply hammerblade/torch/kernel/kernel_threshold.cpp]: Done.

**Comment by yodada** [comment]
Have you tried building cosim? If so we can merge this @zzy82518996

**Review by yodada — APPROVED** [review]
Looks good. Just make sure cosim builds and we are good to go

**Comment by zzy82518996** [comment]
Cosim build passed and minor change on the annotations of some cuda-lite kernels.

**Comment by yodada** [comment]
Sweet! Please go ahead and do the merge once Travis is done ;)

File: .gitignore [hunk 9]
```diff
@@ -271,3 +271,6 @@ DVEfiles
 infinite_mem_stats.csv
 stack.info.*/
 hbKernelCallLog.json
+
+# sinkhorn data
+hammerblade/torch/tests/sinkhorn_tests/sinkhorn_app/data/*
```
Comment (neiladit) [inline .gitignore:276]: This folder is not present anymore so redundant?

**Review by neiladit — APPROVED** [review]
Looks good to me!

**Pull request merged**